pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -874,19 +875,32 @@ protected void doTakeSnapshot() {
         numConsumingSegments++;
         continue;
       }
-      ImmutableSegmentImpl immutableSegment = (ImmutableSegmentImpl) segment;
-      if (!immutableSegment.hasValidDocIdsSnapshotFile()) {
-        segmentsWithoutSnapshot.add(immutableSegment);
+      if (!_updatedSegmentsSinceLastSnapshot.contains(segment)) {
```

Comment:
If we enableSnapshot on an existing upsert table and the set is empty, I think we would skip those existing segments until they get updated or possibly some of them never get updated in long future and they won't have snapshot on disk. (This was a bug I hit on while testing the `_updatedSegmentsSinceLastRefresh`)

We would need to take snapshot for all tracked segments for the first time when to take snapshot for the partition, and then start to skip segments not having updates with this Set.

Reply:
TBH i started this implementation with the exact thing in mind and kept a set of `snapshottedSegments`. But then I realised during server restarts, `addSegment` would put all the segments in `_updatedSegmentsSinceLastSnapshot` right? Enabling snapshotting would require server-restarts as well I suppose?

Reply:
I see. That sounds right to me.